README.md
# YandexTranslator

A simple Elixir interface to Yandex Translate's translation API

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `yandex_translator` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:yandex_translator, "~> 0.9.6"}
  ]
end
```

## Get access to API

### Getting a Cloud Key and Folder ID (for cloud API)

To sign up go [here](https://cloud.yandex.com/docs/translate/concepts/auth)

### Getting a Subscription Key (for old API)

To sign up go [here](https://translate.yandex.ru/developers/keys)

## Configuration

The default behaviour is to configure using the application environment:

In config/config.exs, add:

```elixir
  # for configuring access to cloud api
  config :yandex_translator, cloud_api_key: "API_KEY"
  config :yandex_translator, cloud_folder_id: "FOLDER_ID"

  # for configuring access to old api
  config :yandex_translator, api_key: "API_KEY"
```

## Usage

### Get IAM-token for cloud API access

Request for getting IAM-token for access to cloud API.

```elixir
  # without params
  YandexTranslator.get_iam_token

  # or with key
  YandexTranslator.get_iam_token([key: "API_KEY"])
```

#### Options

    key - API KEY, required or optional (if presented in configuration)

### Supported languages

Request for getting list of supported languages is #langs.
For using cloud api options must contain iam_token param.

```elixir
  # cloud api request
  YandexTranslator.langs([iam_token: ""])

  # old api request
  YandexTranslator.langs([])
```

#### Options for cloud api

    iam_token - IAM-token, required
    folder_id - folder ID of your account at Yandex.Cloud, required or optional (if presented in configuration)

#### Options for old api

    key - access key, required or optional (if presented in config)
    format - one of the [xml|json], default - xml, optional
    ui - language code for getting language translations, optional, example - "en"

### Detection

Request for detecting language of text is #detect.
For using cloud api options must contain iam_token param.

```elixir
  # cloud api request
  YandexTranslator.detect([iam_token: "", text: "Hello"])

  # old api request
  YandexTranslator.detect([text: "Hello"])
```

#### Options for cloud api

    iam_token - IAM-token, required
    folder_id - folder ID of your account at Yandex.Cloud, required or optional (if presented in configuration)
    text - text for detection, required
    hint - list of possible languages, optional, example - "en,ru"

#### Options for old api

    key - access key, required or optional (if presented in config)
    format - one of the [xml|json], default - xml, optional
    text - text, required
    hint - list of possible languages, optional, example - "en,ru"

### Translation

Request for translating text is #translate.
For using cloud api options must contain iam_token param.

```elixir
  # cloud api request
  YandexTranslator.translate([iam_token: "", text: "Hello", target: "ru"])

  # old api request
  YandexTranslator.translate([text: "Hello", lang: "en-ru"])
```
#### Options for cloud api

    iam_token - IAM-token, required
    folder_id - folder ID of your account at Yandex.Cloud, required or optional (if presented in configuration)
    text - text for detection, required
    source - source language, ISO 639-1 format (like "en"), optional
    target - target language, ISO 639-1 format (like "ru"), required
    format - text format, one of the [plain|html], default - plain, optional

#### Options for old api

    key - access key, required or optional (if presented in config)
    format - one of the [xml|json], default - xml, optional
    text - text, required
    lang - direction of translation, required, example - "from-to" or "to"

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/kortirso/yandex_translator.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

## Disclaimer

Use this package at your own peril and risk.

## Documentation

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/yandex_translator](https://hexdocs.pm/yandex_translator).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.9.6] - 2019-01-30
### Fixed
- README

## [0.9.5] - 2019-01-24
### Modified
- using Stream to increase performance

### Fixed
- tests

## [0.9.4] - 2018-12-30
### Modified
- bug with url unencoded

## [0.9.3] - 2018-12-22
### Added
- getting IAM-token for cloud api access
- YandexTranslator.Cloud module for requests to cloud api
- tests

### Modified
- README

## [0.9.2] - 2018-12-21
### Added
- changelog file
- using different api versions

### Modified
- code refactoring
- tests


test/yandex_translator_test.exs
defmodule YandexTranslatorTest do
  use ExUnit.Case

  setup_all do
    {:ok, %{"iamToken" => iam_token}} = YandexTranslator.get_iam_token
    {:ok, iam_token: iam_token}
  end

  describe "cloud api" do
    test "request for getting languages with invalid key" do
      {:error, %{"error_message" => error_message}} = YandexTranslator.langs([iam_token: "1"])

      assert error_message == "rpc error: code = Unauthenticated desc = invalid auth data"
    end

    test "request for detecting language with invalid key" do
      {:error, %{"error_message" => error_message}} = YandexTranslator.detect([iam_token: "1", text: "Hello"])

      assert error_message == "rpc error: code = Unauthenticated desc = invalid auth data"
    end

    test "request for translating text with invalid key" do
      {:error, %{"error_message" => error_message}} = YandexTranslator.translate([iam_token: "1", text: "Hello", lang: "ru"])

      assert error_message == "rpc error: code = Unauthenticated desc = invalid auth data"
    end

    test "request for getting languages", state do
      {:ok, %{"languages" => languages}} = YandexTranslator.langs([iam_token: state[:iam_token]])

      assert is_list(languages) == true
    end


    test "request for detecting language", state do
      {:ok, %{"language" => language}} = YandexTranslator.detect([iam_token: state[:iam_token], text: "Hello"])

      assert language == "en"
    end

    test "request for translating text", state do
      {:ok, %{"translations" => [%{"text" => text}]}} = YandexTranslator.translate([iam_token: state[:iam_token], text: "Hello", target: "es"])

      assert text == "Saludar"
    end
  end

  describe "old api, invalid requests" do
    test "request for getting languages with invalid key" do
      {:error, %{"code" => code, "message" => message}} = YandexTranslator.langs([key: "", format: "json"])

      assert code == 401
      assert message == "API key is invalid"
    end

    test "request for getting languages with invalid key, xml format" do
      {:error, body} = YandexTranslator.langs([key: ""])

      assert String.length(body) != 0
    end

    test "request for detecting language with invalid key" do
      {:error, %{"code" => code, "message" => message}} = YandexTranslator.detect([key: "", text: "Hello", format: "json"])

      assert code == 401
      assert message == "API key is invalid"
    end

    test "request for detecting language with invalid key, xml format" do
      {:error, body} = YandexTranslator.detect([key: "", text: "Hello"])

      assert String.length(body) != 0
    end

    test "request for translating text with invalid key" do
      {:error, %{"code" => code, "message" => message}} = YandexTranslator.translate([key: "", text: "Hello", lang: "ru", format: "json"])

      assert code == 401
      assert message == "API key is invalid"
    end

    test "request for translating text with invalid key, xml format" do
      {:error, body} = YandexTranslator.translate([key: "", text: "Hello", lang: "ru"])

      assert String.length(body) != 0
    end
  end

  describe "old api, valid requests" do
    test "request for getting languages" do
      {:ok, %{"dirs" => dirs}} = YandexTranslator.langs([format: "json"])

      assert is_list(dirs) == true
    end

    test "request for getting languages, xml format" do
      {:ok, body} = YandexTranslator.langs

      assert String.length(body) != 0
    end

    test "request for detecting language" do
      {:ok, %{"code" => code, "lang" => lang}} = YandexTranslator.detect([text: "Hello", format: "json"])

      assert code == 200
      assert lang == "en"
    end

    test "request for detecting language, xml format" do
      {:ok, body} = YandexTranslator.detect([text: "Hello"])

      assert String.length(body) != 0
    end

    test "request for translating text" do
      {:ok, %{"code" => code, "lang" => lang, "text" => text}} = YandexTranslator.translate([text: "Hello", lang: "ru", format: "json"])

      assert code == 200
      assert lang == "en-ru"
      assert text == ["Привет"]
    end

    test "request for translating text, xml format" do
      {:ok, body} = YandexTranslator.translate([text: "Hello", lang: "ru"])

      assert String.length(body) != 0
    end
  end
end


mix.exs
defmodule YandexTranslator.Mixfile do
  use Mix.Project

  @description """
    Translate word and phrases using the Yandex Translate API. Requires that you have a Client ID. See README.md for information.
  """

  def project do
    [
      app: :yandex_translator,
      version: "0.9.6",
      elixir: "~> 1.7",
      description: @description,
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      package: package(),
      deps: deps(),
      source_url: "https://github.com/kortirso/yandex_translator"
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:httpoison, "~> 1.4"},
      {:poison, "~> 3.1.0"},
      {:ex_doc, "~> 0.19", only: :dev}
    ]
  end

  defp package do
    [
      maintainers: ["Bogdanov Anton"],
      licenses: ["MIT"],
      links: %{"github" => "https://github.com/kortirso/yandex_translator"}
    ]
  end
end


lib/yandex_translator/cloud.ex
defmodule YandexTranslator.Cloud do
  @moduledoc """
  Client requests for new version of cloud api
  """

  @iam_token_url "https://iam.api.cloud.yandex.net/iam/v1/tokens"
  @cloud_base_url "https://translate.api.cloud.yandex.net/translate/v1/"
  @iam_token_headers [{"Content-Type", "application/json"}]

  @doc """
  Get IAM-token

  ## Examples

      iex> YandexTranslator.Cloud.get_iam_token([])
      {:ok, %{"iamToken" => ""}}

  """
  @spec get_iam_token(keyword()) :: {:ok, %{iamToken: String.t()}}

  def get_iam_token(args) when is_list(args) do
    args[:key]
    |> fetch_iam_token()
    |> parse()
  end

  defp fetch_iam_token(api_key) do
    body = Poison.encode!(%{"yandexPassportOauthToken" => cloud_api_key(api_key)})
    case HTTPoison.post(@iam_token_url, body, @iam_token_headers) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} -> {:ok, body}
      {:ok, %HTTPoison.Response{body: body}} -> {:error, body}
      {:error, %HTTPoison.Error{reason: reason}} -> {:error, reason}
    end
  end

  defp cloud_api_key(api_key), do: api_key || Application.get_env(:yandex_translator, :cloud_api_key) || ""

  @doc """
  Performs a request

  ## Examples

      iex> YandexTranslator.Cloud.call("languages", args)

  """
  @spec call(String.t(), list) :: {:ok, %{}}

  def call(type, args) when type in ["languages", "detect", "translate"] and is_list(args) do
    type
    |> fetch(args)
    |> parse()
  end

  # MAIN FUNCTIONS

  # make request
  defp fetch(type, args) do
    case HTTPoison.post(@cloud_base_url <> type, body(args, type), headers(args[:iam_token])) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} -> {:ok, body}
      {:ok, %HTTPoison.Response{body: body}} -> {:error, body}
      {:error, %HTTPoison.Error{reason: reason}} -> {:error, reason}
    end
  end

  # parse results
  defp parse({result, response}), do: {result, Poison.Parser.parse!(response)}

  # ADDITIONAL FUNCTIONS

  # generate body for request
  defp body(args, type) do
    args
    |> Keyword.put(:folderId, cloud_folder_id(args[:folder_id]))
    |> Stream.filter(fn {key, _} -> Enum.member?(valid_args(type), key) end)
    |> Stream.map(fn {key, value} -> "#{key}=#{URI.encode_www_form(value)}" end)
    |> Enum.join("&")
  end

  # list with available options based on request type
  defp valid_args(type) do
    case type do
      "languages" -> [:folderId]
      "detect" -> [:folderId, :text, :hint]
      "translate" -> [:folderId, :text, :source, :target, :format]
      _ -> []
    end
  end

  defp cloud_folder_id(nil), do: Application.get_env(:yandex_translator, :cloud_folder_id) || ""
  defp cloud_folder_id(folder_id), do: folder_id

  # define headers for request
  defp headers(iam_token), do: [{"Content-Type", "application/x-www-form-urlencoded"}, {"Authorization", "Bearer #{iam_token}"}]
end


lib/yandex_translator/client.ex
defmodule YandexTranslator.Client do
  @moduledoc """
  Client requests for old version of api, v1.5
  """

  @type api_key :: {:api_key, String.t()}
  @type path :: String.t()

  @doc """
  Performs a request

  ## Examples

      iex> YandexTranslator.Client.call("langs", args)

  """
  @spec call(String.t(), list) :: {}

  def call(type, args) when type in ["langs", "detect", "translate"] and is_list(args) do
    type
    |> generate_url(args)
    |> fetch(args[:format])
    |> parse(args[:format])
  end

  # MAIN FUNCTIONS

  # generate url with all params
  defp generate_url(type, args) do
    args
    |> Keyword.put_new(:key, api_key())
    |> Stream.filter(fn {key, _} -> Enum.member?(valid_args(type), key) end)
    |> Stream.map(fn {key, value} -> "#{key}=#{URI.encode_www_form(value)}" end)
    |> Enum.join("&")
    |> prepare_url(type)
  end

  defp fetch(url, format) do
    case HTTPoison.post(base_url(format) <> url, "", []) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} -> {:ok, body}
      {:ok, %HTTPoison.Response{body: body}} -> {:error, body}
      {:error, %HTTPoison.Error{reason: reason}} -> {:error, reason}
    end
  end

  defp parse({result, response}, "json"), do: {result, Poison.Parser.parse!(response)}
  defp parse(response, _), do: response

  # ADDITIONAL FUNCTIONS

  # list with available options based on request type
  defp valid_args(type) do
    case type do
      "langs" -> [:key, :ui]
      "detect" -> [:key, :text, :hint]
      "translate" -> [:key, :text, :lang]
      _ -> []
    end
  end

  # add type of request to url
  defp prepare_url(url, type) do
    case type do
      "langs" -> "/getLangs?#{url}"
      "detect" -> "/detect?#{url}"
      "translate" -> "/translate?#{url}"
      _ -> ""
    end
  end

  # define params for request
  defp base_url("json"), do: "https://translate.yandex.net/api/v1.5/tr.json"
  defp base_url(_), do: "https://translate.yandex.net/api/v1.5/tr"
  
  # API KEY from config
  defp api_key, do: Application.get_env(:yandex_translator, :api_key) || ""
end


lib/yandex_translator.ex
defmodule YandexTranslator do
  @moduledoc """
  Elixir client for Yandex.Translate API

  ## Configuration (new API)
  An API key and folder id can be set in your application's config.
  For getting api key and folder is check readme.

      config :yandex_translator, cloud_api_key: "API_KEY"
      config :yandex_translator, cloud_folder_id: "FOLDER_ID"

  ## Configuration (old API)
  An API key can be set in your application's config.
  For getting api key check readme.

      config :yandex_translator, api_key: "API_KEY"

  """

  alias YandexTranslator.{Client, Cloud}

  @doc """
  Get IAM-token for using it in requests to Yandex.Cloud
  Valid 12 hours.

  ## Example

      iex> YandexTranslator.get_iam_token
      {:ok, %{"iamToken" => ""}}

  """
  @spec get_iam_token() :: {:ok, %{iamToken: String.t()}}

  def get_iam_token, do: get_iam_token([])

  @doc """
  Get IAM-token for using it in requests to Yandex.Cloud
  Valid 12 hours.

  ## Example

      iex> YandexTranslator.get_iam_token([])
      {:ok, %{"iamToken" => ""}}

  ### Options

      key - API KEY, required or optional (if presented in configuration)

  """
  @spec get_iam_token(keyword()) :: {:ok, %{iamToken: String.t()}}

  def get_iam_token(options) when is_list(options), do: Cloud.get_iam_token(options)

  @doc """
  Get available languages for translation

  ## Example

      iex> YandexTranslator.langs

  """
  @spec langs() :: {:ok, %{}}

  def langs, do: langs([])

  @doc """
  Get available languages for translation
  For using cloud api options must contain iam_token param.

  ## Example (cloud API)

      iex> YandexTranslator.langs([iam_token: ""])
      {:ok, %{"languages" => [%{"language" => "az"}, %{...}, ...]}}

  ### Options (cloud API)

      iam_token - IAM-token, required
      folder_id - folder ID of your account at Yandex.Cloud, required or optional (if presented in configuration)

  ## Example (old API)

      iex> YandexTranslator.langs([format: "json"])
      {:ok, %{"dirs" => ["az-ru", ...]}}

  ### Options (old API)

      key - API KEY, required or optional (if presented in configuration)
      format - one of the [xml|json], optional, default - xml
      ui - language code for getting language translations, optional, example - "en"

  """
  @spec langs(keyword()) :: {:ok, %{}}

  def langs(options) when is_list(options) do
    case options[:iam_token] do
      nil -> Client.call("langs", options)
      _ -> Cloud.call("languages", options)
    end
  end

  @doc """
  Detect language for text
  For using cloud api options must contain iam_token param.

  ## Example (cloud API)

      iex> YandexTranslator.detect([iam_token: "", text: "Hello"])
      {:ok, %{"language" => "en"}}

  ### Options (cloud API)

      iam_token - IAM-token, required
      folder_id - folder ID of your account at Yandex.Cloud, required or optional (if presented in configuration)
      text - text for detection, required
      hint - list of possible languages, optional, example - "en,ru"

  ## Example (old API)

      iex> YandexTranslator.detect([text: "Hello", format: "json"])
      {:ok, %{"code" => 200, "lang" => "en"}}

  ## Options (old API)

      key - API KEY, required or optional (if presented in config)
      format - one of the [xml|json], optional, default - xml
      text - text for detection, required
      hint - list of possible languages, optional, example - "en,ru"

  """
  @spec detect(keyword()) :: {:ok, %{}}

  def detect(options) when is_list(options) do
    case options[:iam_token] do
      nil -> Client.call("detect", options)
      _ -> Cloud.call("detect", options)
    end
  end

  @doc """
  Translate word or phrase
  For using cloud api options must contain iam_token param.

  ## Example (cloud API)

      iex> YandexTranslator.translate([iam_token: iam_token, text: "hello world", source: "en", target: "es"])
      {:ok, %{"translations" => [%{"text" => "hola mundo"}]}}

  ### Options (cloud API)

      iam_token - IAM-token, required
      folder_id - folder ID of your account at Yandex.Cloud, required or optional (if presented in configuration)
      text - text for detection, required
      source - source language, ISO 639-1 format (like "en"), optional
      target - target language, ISO 639-1 format (like "ru"), required
      format - text format, one of the [plain|html], default - plain, optional

  ## Example (old API)

      iex> YandexTranslator.translate([format: "json", text: "Hello", lang: "en-es"])
      {:ok, %{"code" => 200, "lang" => "en-es", "text" => ["Hola"]}}

  ## Options (old API)

      key - API KEY, required or optional (if presented in config)
      format - one of the [xml|json], optional, default - xml
      text - text, required
      lang - direction of translation, optional, example - "from-to" or "to"

  """
  @spec translate(keyword()) :: {:ok, %{}}

  def translate(options) when is_list(options) do
    case options[:iam_token] do
      nil -> Client.call("translate", options)
      _ -> Cloud.call("translate", options)
    end
  end
end
